Change Neo3Boa Documentation describing the features (first iteration)

luc10921 opened the pull request:
**Summary or solution description**
- Removed repetitions on README, CONTRIBUTING and docs;
- Added execution examples on most functions and classes (some like ByteString and Neo3.6 methods were skipped);
- Added some interops that were missing on the docs;
- Removed subpackages on interops, since the classes are now shown on the package module;
- Changed some files from `rst` to `md`.

lllwvlvwlll commented:
Task linked: [CU-864ej3bwe Change Neo3Boa Documentation describing the features (first iteration)](https://app.clickup.com/t/864ej3bwe)

melanke commented on docs/source/calling-smart-contracts.md:
```diff
@@ -0,0 +1,125 @@
+Instead of manually writing the smart contract interface, you can use [CPM](https://github.com/CityOfZion/cpm/tree/master#readme) 
+to generate it automatically. After installing Neo3-boa, you can install CPM by typing `install_cpm` on CLI (without the 
+`neo3-boa` prefix). Then, you'll need to create a [cpm.yaml config file](https://github.com/CityOfZion/cpm/blob/master/docs/config.md) 
+and put the smart contract information there, and [run cpm](https://github.com/CityOfZion/cpm#example-commands).
```
```suggestion
, put the smart contract information there, and [run cpm](https://github.com/CityOfZion/cpm#example-commands).
```

melanke commented on docs/source/getting-started.md:
```diff
@@ -0,0 +1,110 @@
+
+| Packages                                                                                                                             | Contains:                                                                                                                               | Important features                                                                                                           ...
+|--------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------- ...
+| [boa3.builtin](https://dojo.coz.io/neo3/boa/boa3/builtin/boa3-builtins.html )                                                        | all other packages, it also contains an env variable that lets you change the value of a variable when compiling the smart contract.    | [env]()                                                                                                                      ...
```
`all other packages` is strange when this is the first item in the list

melanke commented on docs/source/getting-started.md:
```diff
@@ -0,0 +1,110 @@
+| Packages                                                                                                                             | Contains:                                                                                                                               | Important features                                                                                                           ...
+|--------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------- ...
+| [boa3.builtin](https://dojo.coz.io/neo3/boa/boa3/builtin/boa3-builtins.html )                                                        | all other packages, it also contains an env variable that lets you change the value of a variable when compiling the smart contract.    | [env]()                                                                                                                      ...
+| [boa3.builtin.compile_time](https://dojo.coz.io/neo3/boa/boa3/builtin/compile-time/boa3-builtin-compile-time.html)                   | methods and classes that are need when you are compiling your smart contract, as opposed to when it's being executed on the blockchain. | [public](https://dojo.coz.io/neo3/boa/boa3/builtin/compile-time/boa3-builtin-compile-time.html#boa3.builtin.compile_time.pub ...
```
methods and classes that are **needed** when

melanke commented on docs/source/getting-started.md:
```diff
@@ -0,0 +1,110 @@
+| [boa3.builtin.interop.iterator](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/iterator/boa3-builtin-interop-iterator.html)       | the iterator class.                                                                                                                     | [Iterator](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/iterator/boa3-builtin-interop-iterator.html#boa3.builtin.intero ...
+| [boa3.builtin.interop.json](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/json/boa3-builtin-interop-json.html)                   | methods to serialize and deserialize JSON.                                                                                              | [json_serialize](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/json/boa3-builtin-interop-json.html#boa3.builtin.interop. ...
+| [boa3.builtin.interop.oracle](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/oracle/boa3-builtin-interop-oracle.html)             | features related with Neo Oracle, it is used to get information from outside the blockchain.                                            | [Oracle](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/oracle/boa3-builtin-interop-oracle.html#boa3.builtin.interop.orac ...
+| [boa3.builtin.interop.policy](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/policy/boa3-builtin-interop-policy.html)             | features related with the policy that affects all of the Neo blockchain.                                                                | [get_exec_fee_factor](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/policy/boa3-builtin-interop-policy.html#boa3.builtin ...
```
features related to policies that affect the entire Neo blockchain.

melanke commented on docs/source/getting-started.md:
````diff
@@ -0,0 +1,110 @@
+    return storage.get("first script")              # the get method will return the value associated with "first script" key
+```
+
+Alternatively, instead of creating a `save_hello_world` function and having to call it yourself, you could instead create
````
I don't think we should say it is "Alternatively", because you can do both on the same contract. Let's use this as a seconds step of the tutorial with a new title

melanke commented on docs/source/getting-started.md:
```diff
@@ -0,0 +1,110 @@
+
+## What's next
+- [Testing and debugging your smart contract](./testing-and-debugging.md)
+- [How to call other smart contracts on the blockchain](./calling-smart-contracts.md)
```
let's discuss what other steps we should add

melanke commented on docs/source/testing-and-debugging.md:
````diff
@@ -0,0 +1,99 @@
+    runner.execute()
+    assert invoke.result is None
+```
+
````
We also need to point how the user will interact with the SmartContract from outside, in a real world scenario, not only unit tests, but an application. Let's discuss this

melanke reviewed (changes requested):
The first iteration of review, I think there will be a few more

meevee98 commented on docs/source/getting-started.md:
```diff
@@ -0,0 +1,110 @@
+| [boa3.builtin.interop.storage](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/storage/boa3-builtin-interop-storage.html)          | features to store, get, or change values inside the blockchain.                                                                         | [get](https://dojo.coz.io/neo3/boa/boa3/builtin/interop/storage/boa3-builtin-interop-storage.html#boa3.builtin.interop.stora ...
+| [boa3.builtin.nativecontract](https://dojo.coz.io/neo3/boa/boa3/builtin/nativecontract/boa3-builtin-nativecontract.html)             | classes that interface Neo's native contracts.                                                                                          | [ContractManagement](https://dojo.coz.io/neo3/boa/boa3/builtin/nativecontract/boa3-builtin-nativecontract-contractmanagement ...
+| [boa3.builtin.type](https://dojo.coz.io/neo3/boa/boa3/builtin/type/boa3-builtin-type.html)                                           | Neo types.                                                                                                                              | [UInt160](https://dojo.coz.io/neo3/boa/boa3/builtin/type/boa3-builtin-type.html#boa3.builtin.type.UInt160), [UInt256](https: ...
+| [boa3.builtin.vm](https://dojo.coz.io/neo3/boa/boa3/builtin/vm/boa3-builtin-vm.html)                                                 | Opcodes used internally by the NeoVM, you'll probably won't need to use this package.                                                   | [Opcode](https://dojo.coz.io/neo3/boa/boa3/builtin/vm/boa3-builtin-vm.html#boa3.builtin.vm.Opcode)                           ...
```
Opcodes used internally by the Neo VM, used to create scripts dynamically.

meevee98 commented on docs/source/getting-started.md:
````diff
@@ -0,0 +1,110 @@
+### Using CLI
+
+```shell
+$ neo3-boa path/to/your/file.py
````
we need to ensure that we change here to the following when we release the new documentation.
```suggestion
$ neo3-boa compile path/to/your/file.py
```
On master branch and release versions we still use `neo3-boa` command, but on development branch `neo3-boa compile` is already implemented

meevee98 commented on docs/source/testing-and-debugging.md:
````diff
@@ -0,0 +1,99 @@
+### Using CLI
+
+```shell
+$ neo3-boa path/to/your/file.py -d|--debug
````
same here, we need to ensure the documentation is consistent with each branch
```suggestion
$ neo3-boa compile path/to/your/file.py -d|--debug
```

meevee98 commented on docs/source/testing-and-debugging.md:
````diff
@@ -0,0 +1,99 @@
+
+Install [Neo-Express](https://github.com/neo-project/neo-express#neo-express-and-neo-trace) and [Neo Test Runner](https://github.com/ngdenterprise/neo-test#neo-test-runner).
+
+```shell
````
We should refer to each tool installation page instead of showing the commands because the installation varies in each operational system

meevee98 commented on docs/source/calling-smart-contracts.md:
```diff
@@ -0,0 +1,125 @@
+## with call_contract
+Use the `call_contract` method from `boa3.builtin.interop.contract` on your smart contract. You'll need to use the 
+script hash, followed by the name of the function you want to call, followed by the arguments of said function. You'll 
+also need to type cast the return, otherwise it will be considered as `Any`.
```
You'll also need to type cast the return **so the compiler type checker works as expected**, otherwise the return type will be considered as `Any`

meevee98 commented on docs/source/calling-smart-contracts.md:
```diff
@@ -0,0 +1,125 @@
+@public
+def calling_other_contract() -> str:
+    d6_roll = Dice.rand_between(1, 6)
+    return "Die result is " + str(d6_roll)
```
```suggestion
    return "Dice result is " + str(d6_roll)
```
also it should be good to use interop and native contract methods in the examples to better example how Neo developing works (not mandatory in this PR)
```suggestion
    return "Dice result is " + StdLib.itoa(d6_roll)
```

meevee98 commented on boa3/builtin/type/__init__.py:
```diff
@@ -20,6 +20,15 @@
 class Event:
     """
     Describes an action that happened in the blockchain.
+
+    >>> from boa3.builtin.compile_time import CreateNewEvent
+    ... new_event: Event = CreateNewEvent(  # create a new Event with the CreateNewEvent method
+    ...     [
+    ...        ('name', str),
+    ...        ('amount', int)
+    ...     ],
+    ...     'New Event'
+    ... )
```
```suggestion

    Neo3-boa compiler won't recognize the __init__ of this class. To create a new Event, with the method `CreateNewEvent`:
    
    >>> from boa3.builtin.compile_time import CreateNewEvent
    ... new_event: Event = CreateNewEvent(  # create a new Event with the CreateNewEvent method
    ...     [
    ...        ('name', str),
    ...        ('amount', int)
    ...     ],
    ...     'New Event'
    ... )
```

luc10921 commented:
I had to rebase to get the newest ByteString changes, on the new commit I:
- implemented most feedbacks;
  - The one I didn't do was : https://github.com/CityOfZion/neo3-boa/pull/1061#discussion_r1218256697
- removed all warnings that would show up when building the docs;
- stopped autogenerating private methods on docs;
- added a section talking about how to invoke your deployed smart contract with NeoNova;
- added interop subpackages into the main package instead of having a page for each one (for example, 'Block' and 'Transaction' classes are inside the 'interop.blockchain' page).
- changed the links on the markdown, because they were referencing the dojo.coz.io documentation that might not have the method/class/variable/same_path on the newest version

melanke commented on boa3/builtin/compile_time/__init__.py:
```diff
@@ -53,6 +104,18 @@ def contract(script_hash: Union[str, bytes]):
     """
     This decorator identifies a class that should be interpreted as an interface to an existing contract.
 
+    Check out `Our Documentation <https://dojo.coz.io/neo3/boa/calling-smart-contracts.html#with-interface>`__ to learn
```
I think this link is broken

melanke commented on docs/source/invoking-with-neonova.md:
```diff
@@ -0,0 +1,28 @@
+input your local address with the rpc-port and magic number (they should be inside the .neo-express file), and click on
+"Save".
+
+![Setting Local Network](./images/InvokeNeoNova_1Network.gif)
```
We should use Neon Wallet Desktop instead of NeoLine

melanke commented on docs/source/invoking-with-neonova.md:
```diff
@@ -0,0 +1,28 @@
+
+![Adding Wallet](./images/InvokeNeoNova_2Wallet.gif)
+
+Then, select the "Read" or "Write" type on the menu depending on what you want to do, input the script hash of the smart 
```
> depending on what you want to do

It would be interesting to explain the differences and that you can "test" your "write" invocations, etc.
